src/wasm/wasm-type.cpp

```diff
@@ -1232,6 +1242,41 @@ HeapType HeapType::getLeastUpperBound(HeapType a, HeapType b) {
   return TypeBounder().getLeastUpperBound(a, b);
 }
 
+// Recursion groups with single elements are encoded as that single element's
+// type ID with the low bit set and other recursion groups are encoded with the
+// address of the vector containing their members. These encodings are disjoint
+// because the alignment of the vectors is greater than 1.
+static_assert(alignof(std::vector<HeapType>) > 1);
```

Comment:
If the case of singleton rec groups is rare (again, I assume a single big rec group will be most common), maybe it's not important to optimize for this? That is, we could have normal rec groups for singletons.